src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs

```diff
@@ -45,60 +57,224 @@ internal partial class CommonReferenceManager<TCompilation, TAssemblySymbol>
 
-            hasCircularReference = CheckCircularReference(boundInputs);
+            // tracks identities we already asked the resolver to resolve:
+            var resolvedMissingIdentities = PooledHashSet<AssemblyIdentity>.GetInstance();
 
-            corLibraryIndex = IndexOfCorLibrary(assemblies);
+            // reference bindings of implicit assemblies, use to calculate a fixed point:
+            var referenceBindingsToProcess = ArrayBuilder<AssemblyReferenceBinding[]>.GetInstance();
 
-            // For each assembly, locate AssemblySymbol with similar reference resolution
-            // What does similar mean?
-            // Similar means: 
-            // 1) The same references are resolved against the assemblies that we are given.
-            // 2) The same assembly is used as the COR library.
+            try
+            {
+                // collect all missing identities, resolve the assemblies and bind their references against explicit definitions:
+                referenceBindingsToProcess.AddRange(referenceBindings);
 
-            TAssemblySymbol[] candidateInputAssemblySymbols = new TAssemblySymbol[totalAssemblies];
+                while (referenceBindingsToProcess.Count > 0)
+                {
+                    foreach (var binding in referenceBindingsToProcess.Pop())
+                    {
+                        // try resolve a reference if eitehr the binding failed or didn't find an exact match:
+                        if (binding.IsBound && binding.VersionDifference == 0 || 
+                            resolvedMissingIdentities.Contains(binding.ReferenceIdentity))
+                        {
+                            continue;
+                        }
 
-            // If any assembly from assemblies array refers back to assemblyBeingBuilt,
-            // we know that we cannot reuse symbols for any assemblies containing NoPia
-            // local types. Because we cannot reuse symbols for assembly referring back
-            // to assemblyBeingBuilt.
-            if (!hasCircularReference)
-            {
-                // Deal with assemblies containing NoPia local types.
-                if (ReuseAssemblySymbolsWithNoPiaLocalTypes(boundInputs, candidateInputAssemblySymbols, assemblies, corLibraryIndex))
+                        var peReference = resolver.ResolveMissingAssembly(binding.ReferenceIdentity);
+                        if (peReference == null)
+                        {
+                            continue;
+                        }
+
+                        var data = ResolveMissingAssembly(binding.ReferenceIdentity, peReference, importOptions, resolutionDiagnostics);
+                        if (data == null)
+                        {
+                            continue;
+                        }
+
+                        resolvedReferences.Add(peReference);
+                        resolvedMissingIdentities.Add(binding.ReferenceIdentity);
```

Comment:
It feels that this should be done right after the ```if``` checking for ```resolvedMissingIdentities.Contains(binding.ReferenceIdentity)```. Otherwise we might attempt to resolve the same reference over and over again if we ```continue``` above.